sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPluginInternal.scala

```diff
@@ -640,11 +640,36 @@ object ScalaJSPluginInternal {
       loadedJSEnv := {
         val log = streams.value.log
         val env = resolvedJSEnv.value
-        val libs =
+        val deps =
           resolvedJSDependencies.value.data ++ scalaJSConfigurationLibs.value
+
+        /* Implement the behavior of commonJSName without having to burn it
+         * inside NodeJSEnv, and hence in the JSEnv API.
+         * Since this matches against NodeJSEnv specifically, it obviously
+         * breaks the OO approach, but oh well ...
+         */
```

Comment:
Yes, indeed, I think they are mostly independent. Or rather, I think that `CommonJSModule` does not preclude a browser environment, as long as it is used with an appropriate bundler (e.g., webpack in scalajs-bundler). Using `NoModule` but actually running on Node.js (in prod, I mean) should not happen, although there probably still are occurrences of that in codebases created before the introduction of `CommonJSModule`. Your list of facades actually provides one such example: the AWS facade.

The Phaser facade is also interesting. It is actually a very good example of why `commonJSName` and more generally `jsDependencies` is dangerously misdesigned: some classes such as [`pixi.Sprite`](https://github.com/dpitkevics/Phaser-Scala-JS/blob/master/src/main/scala/org/scalajs/phaser/pixi/Sprite.scala) actually have an `@JSName("PIXI.Sprite")`, referring to a different top-level JS object than the `commonJSName` they use. That class will work in the browser, but not in Node.js.

> However, I still do not understand what your desired end-state with respect to commonJSName is

In my mind, `commonJSName` as a surface feature is tight to `jsDependencies`, as it was added as a patch because `jsDependencies` did not work for "too smart" libraries. I would therefore move it together with `jsDependencies` in a separate plugin. `jsDependencies` itself, I basically consider to be on life support in 1.x. We keep it running as before for people who were using it, but we do not invest in it anymore.

As such, I don't consider the hack of the instance test on `NodeJSEnv` particularly annoying. It's a life-support hack. That said, if people complain, it would be easy to add a configurable predicate instead, so that people have an additional escape hatch if they use some other obscure CommonJS-based runtime.

> Add an option to the new bundler plugin to chose the behavior? (Probably the best option at this point).

Not sure what you are referring to with "the new bundler plugin". If you are talking about scalajs-bundler, it can already do that. It's even in its cookbook: https://scalacenter.github.io/scalajs-bundler/cookbook.html#global-namespace. That recipe basically forces an (npm) library to be exposed as a global object, instead of having to be `require`d.